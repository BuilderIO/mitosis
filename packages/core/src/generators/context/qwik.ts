import { format } from 'prettier/standalone';
import { MitosisContext } from '../../types/mitosis-context';

type ContextToQwikOptions = {
  format?: boolean;
};

export const contextToQwik =
  (options: ContextToQwikOptions = {}) =>
  ({ context }: { context: MitosisContext }): string => {
    let str = `
  import { createContext } from '@builder.io/qwik';

  export default createContext<any>("${context.name}")
  `;

    if (options.format !== false) {
      try {
        str = format(str, {
          parser: 'typescript',
          plugins: [
            require('prettier/parser-typescript'), // To support running in browsers
          ],
        });
      } catch (err) {
        console.error('Format error for file:', str);
        throw err;
      }
    }

    return str;
  };
