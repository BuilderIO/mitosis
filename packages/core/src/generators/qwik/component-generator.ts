import { format } from 'prettier/standalone';
import { fastClone } from '../../helpers/fast-clone';
import { collectCss } from '../../helpers/styles/collect-css';
import { MitosisComponent } from '../../types/mitosis-component';
import { BaseTranspilerOptions, TranspilerGenerator } from '../../types/transpiler';
import { checkHasState } from '../../helpers/state';
import { addPreventDefault } from './helpers/add-prevent-default';
import { renderJSXNodes } from './jsx';
import { arrowFnBlock, File, invoke, SrcBuilder } from './src-generator';
import {
  Plugin,
  runPostCodePlugins,
  runPostJsonPlugins,
  runPreCodePlugins,
  runPreJsonPlugins,
} from '../../modules/plugins';
import { stableInject } from './helpers/stable-inject';
import { mergeOptions } from '../../helpers/merge-options';
import { emitStateMethodsAndRewriteBindings, emitUseStore, StateInit } from './helpers/state';
import { CODE_PROCESSOR_PLUGIN } from '../../helpers/plugins/process-code';
import { replaceIdentifiers, replaceStateIdentifier } from '../../helpers/replace-identifiers';
import { convertTypeScriptToJS } from '../../helpers/babel-transform';

Error.stackTraceLimit = 9999;

const DEBUG = false;

export interface ToQwikOptions extends BaseTranspilerOptions {}

const PLUGINS: Plugin[] = [
  () => ({
    json: {
      post: (json) => {
        addPreventDefault(json);

        return json;
      },
    },
  }),
  CODE_PROCESSOR_PLUGIN((codeType, json) => {
    switch (codeType) {
      case 'bindings':
      case 'state':
      case 'hooks':
      case 'hooks-deps':
      case 'properties':
      case 'dynamic-jsx-elements':
        // update signal getters to have `.value`
        return (code, k) => {
          // `ref` should not update the signal value access
          if (k === 'ref') {
            return code;
          }
          Object.keys(json.refs).forEach((ref) => {
            code = replaceIdentifiers({
              code,
              from: ref,
              to: (x) => (x === ref ? `${x}.value` : `${ref}.value.${x}`),
            });
          });
          // update signal getters to have `.value`
          return replaceStateIdentifier((name) => {
            const state = json.state[name];
            switch (state?.type) {
              case 'getter':
                return `${name}.value`;

              case 'function':
              case 'method':
              case 'property':
              case undefined:
                return `state.${name}`;
            }
          })(code);
        };
    }
  }),
];

const DEFAULT_OPTIONS: ToQwikOptions = {
  plugins: PLUGINS,
};

export const componentToQwik: TranspilerGenerator<ToQwikOptions> =
  (userOptions = {}) =>
  ({ component: _component, path }): string => {
    // Make a copy we can safely mutate, similar to babel's toolchain
    let component = fastClone(_component);

    const options = mergeOptions(DEFAULT_OPTIONS, userOptions);

    component = runPreJsonPlugins(component, options.plugins);
    component = runPostJsonPlugins(component, options.plugins);

    const isTypeScript = !!options.typescript;
    const file = new File(
      component.name + (isTypeScript ? '.ts' : '.js'),
      {
        isPretty: true,
        isJSX: true,
        isTypeScript: isTypeScript,
        isModule: true,
        isBuilder: false,
      },
      '@builder.io/qwik',
      '',
    );
    try {
      emitImports(file, component);
      emitTypes(file, component);
      emitExports(file, component);
      const metadata: Record<string, any> = component.meta.useMetadata || ({} as any);
      const isLightComponent: boolean = metadata?.qwik?.component?.isLight || false;
      const mutable: string[] = metadata?.qwik?.mutable || [];

      const imports: Record<string, string> = metadata?.qwik?.imports || {};
      Object.keys(imports).forEach((key) => file.import(imports[key], key));

      const state: StateInit = emitStateMethodsAndRewriteBindings(file, component, metadata);
      const hasState = checkHasState(component);

      let css: string | null = null;

      const componentFn = arrowFnBlock(
        ['props'],
        [
          function (this: SrcBuilder) {
            css = emitUseStyles(file, component);
            emitUseContext(file, component);
            emitUseRef(file, component);
            hasState &&
              emitUseStore({ file, stateInit: state, isDeep: metadata?.qwik?.hasDeepStore });
            emitUseComputed(file, component);
            emitUseContextProvider(file, component);
            emitUseClientEffect(file, component);
            emitUseMount(file, component);
            emitUseTask(file, component);

            emitJSX(file, component, mutable);
          },
        ],
        [(component.propsTypeRef || 'any') + (isLightComponent ? '&{key?:any}' : '')],
      );
      file.src.const(
        component.name,
        isLightComponent
          ? componentFn
          : invoke(file.import(file.qwikModule, 'component$'), [componentFn]),
        true,
        true,
      );
      file.exportDefault(component.name);
      emitStyles(file, css);
      DEBUG && file.exportConst('COMPONENT', JSON.stringify(component));
      let sourceFile = file.toString();
      sourceFile = runPreCodePlugins(sourceFile, options.plugins);
      sourceFile = runPostCodePlugins(sourceFile, options.plugins);
      return sourceFile;
    } catch (e) {
      console.error(e);
      return (e as Error).stack || String(e);
    }
  };

function emitExports(file: File, component: MitosisComponent) {
  Object.keys(component.exports || {}).forEach((key) => {
    const exportObj = component.exports![key]!;
    const code = exportObj.code.startsWith('export ') ? exportObj.code : `export ${exportObj.code}`;
    file.src.emit(code);
  });
}

function emitUseClientEffect(file: File, component: MitosisComponent) {
  if (component.hooks.onMount) {
    // This is called useMount, but in practice it is used as
    // useClientEffect. Not sure if this is correct, but for now.
    const code = component.hooks.onMount.code;
    file.src.emit(file.import(file.qwikModule, 'useVisibleTask$').localName, '(()=>{', code, '});');
  }
}

function emitUseMount(file: File, component: MitosisComponent) {
  if (component.hooks.onInit) {
    const code = component.hooks.onInit.code;
    file.src.emit(file.import(file.qwikModule, 'useTask$').localName, '(()=>{', code, '});');
  }
}

function emitUseTask(file: File, component: MitosisComponent) {
  if (component.hooks.onUpdate) {
    component.hooks.onUpdate.forEach((onUpdate) => {
      file.src.emit(file.import(file.qwikModule, 'useTask$').localName, '(({track})=>{');
      emitTrackExpressions(file.src, onUpdate.deps);
      file.src.emit(convertTypeScriptToJS(onUpdate.code));
      file.src.emit('});');
    });
  }
}

function emitTrackExpressions(src: SrcBuilder, deps?: string) {
  if (!deps) {
    return;
  }

  const dependencies = deps.substring(1, deps.length - 1).split(',');
  dependencies.forEach((dep) => {
    src.emit(`track(() => ${dep});`);
  });
}

function emitJSX(file: File, component: MitosisComponent, mutable: string[]) {
  const directives = new Map();
  const handlers = new Map<string, string>();
  const styles = new Map();
  const parentSymbolBindings = {};
  if (file.options.isPretty) {
    file.src.emit('\n\n');
  }
  file.src.emit(
    'return ',
    renderJSXNodes(
      file,
      directives,
      handlers,
      component.children,
      styles,
      null,
      parentSymbolBindings,
    ),
  );
}

function emitUseContextProvider(file: File, component: MitosisComponent) {
  Object.entries(component.context.set).forEach(([_ctxKey, context]) => {
    file.src.emit(`
      ${file.import(file.qwikModule, 'useContextProvider').localName}(
        ${context.name}, ${file.import(file.qwikModule, 'useStore').localName}({
      `);

    for (const [prop, propValue] of Object.entries(context.value || {})) {
      file.src.emit(`${prop}: `);
      switch (propValue?.type) {
        case 'getter':
          file.src.emit(`(()=>{
            ${extractGetterBody(propValue.code)}
          })()`);
          break;

        case 'function':
        case 'method':
          throw new Error('Qwik: Functions are not supported in context');

        case 'property':
          file.src.emit(stableInject(propValue.code));
          break;
      }
      file.src.emit(',');
    }
    file.src.emit('}));');
  });
}

function emitUseContext(file: File, component: MitosisComponent) {
  Object.keys(component.context.get).forEach((ctxKey) => {
    const context = component.context.get[ctxKey];
    file.src.emit(
      'const ',
      ctxKey,
      '=',
      file.import(file.qwikModule, 'useContext').localName,
      '(',
      context.name,
      ');',
    );
  });
}

function emitUseRef(file: File, component: MitosisComponent) {
  Object.keys(component.refs).forEach((refKey) => {
    file.src.emit(
      `const `,
      refKey,
      '=',
      file.import(file.qwikModule, 'useSignal').localName,
      `${file.options.isTypeScript ? '<Element>' : ''}();`,
    );
  });
}

function emitUseStyles(file: File, component: MitosisComponent): string {
  const css = collectCss(component, { prefix: component.name });
  if (css) {
    file.src.emit(file.import(file.qwikModule, 'useStylesScoped$').localName, '(STYLES);');
    if (file.options.isPretty) {
      file.src.emit('\n\n');
    }
  }
  return css;
}

function emitStyles(file: File, css: string | null) {
  if (!css) {
    return;
  }

  if (file.options.isPretty) {
    file.src.emit('\n\n');
    try {
      css = format(css, {
        parser: 'css',
        plugins: [
          // To support running in browsers
          require('prettier/parser-postcss'),
        ],
      });
    } catch (e) {
      throw new Error(
        e +
          '\n' +
          '========================================================================\n' +
          css +
          '\n\n========================================================================',
      );
    }
  }
  file.exportConst('STYLES', '`\n' + css.replace(/`/g, '\\`') + '`\n');
}

function emitTypes(file: File, component: MitosisComponent) {
  if (file.options.isTypeScript) {
    component.types?.forEach((t) => file.src.emit(t, '\n'));
  }
}

function emitImports(file: File, component: MitosisComponent) {
  // <SELF> is used for self-referencing within the file.
  file.import('<SELF>', component.name);
  component.imports?.forEach((i) => {
    Object.keys(i.imports).forEach((key) => {
      const keyValue = i.imports[key]!;
      file.import(i.path.replace('.lite', '').replace('.tsx', ''), keyValue, key);
    });
  });
}

function extractGetterBody(code: string): string {
  const start = code.indexOf('{');
  const end = code.lastIndexOf('}');
  return code.substring(start + 1, end).trim();
}

function emitUseComputed(file: File, component: MitosisComponent) {
  for (const [key, stateValue] of Object.entries(component.state)) {
    switch (stateValue?.type) {
      case 'getter':
        file.src.const(`
          ${key} = ${file.import(file.qwikModule, 'useComputed$').localName}(() => {
            ${extractGetterBody(stateValue.code)}
          })
        `);
        continue;
    }
  }
}
