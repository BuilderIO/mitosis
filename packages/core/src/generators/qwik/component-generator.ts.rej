diff a/packages/core/src/generators/qwik/component-generator.ts b/packages/core/src/generators/qwik/component-generator.ts	(rejected hunks)
@@ -25,6 +25,7 @@ import {
   StateInit,
 } from './helpers/state';
 import { convertTypeScriptToJS } from './helpers/transform-code';
+import { CAMEL_CASE_PLUGIN } from 'src/helpers/plugins/map-camel-cased-attributes';
 
 Error.stackTraceLimit = 9999;
 
@@ -42,6 +43,7 @@ const PLUGINS: Plugin[] = [
       },
     },
   }),
+  CAMEL_CASE_PLUGIN
 ];
 
 const DEFAULT_OPTIONS: ToQwikOptions = {
