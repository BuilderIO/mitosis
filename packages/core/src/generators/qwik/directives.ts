import { MitosisNode } from '../../types/mitosis-node';
import { INDENT, SrcBuilder, NL, WS, UNINDENT } from './src-generator';

export const DIRECTIVES: Record<
  string,
  (node: MitosisNode, blockFn: () => void) => void
> = {
  Show: (node: MitosisNode, blockFn: () => void) =>
    function(this: SrcBuilder) {
      const expr = node.bindings.when;
      this.isJSX && this.emit('{', WS);
      this.emit(expr, WS, '?', INDENT, NL);
      blockFn();
      this.emit(':', WS, 'null', UNINDENT, NL);
      this.isJSX && this.emit('}', NL);
    },
  For: (node: MitosisNode, blockFn: () => void) =>
    function(this: SrcBuilder) {
      const expr = node.bindings.each;
      this.isJSX && this.emit('{', WS);
      this.emit('(', expr, WS, '||', WS, '[])');
      this.emit('.forEach(', '()', WS, '=>', WS, '(', INDENT, NL);
      blockFn();
      this.emit(UNINDENT, ')', ')', NL);
      this.isJSX && this.emit('}', NL);
    },
};
