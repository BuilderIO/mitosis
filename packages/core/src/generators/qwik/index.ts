export { addCommonStyles, addComponent, createFileSet } from './component';
export type { FileSet, QwikOptions } from './component';
export { componentToQwik } from './component-generator';
export { File } from './src-generator';
