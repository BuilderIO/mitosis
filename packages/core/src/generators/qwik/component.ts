import { stat } from 'fs';
import {
  compileAwayBuilderComponentsFromTree,
  components as compileAwayComponents,
} from '../../plugins/compile-away-builder-components';
import { MitosisComponent } from '../../types/mitosis-component';
import { minify } from '../minify';
import { renderHandlers } from './handlers';
import { renderJSXNodes } from './jsx';
import {
  arrowFnBlock,
  arrowFnValue,
  File,
  iif,
  INDENT,
  invoke,
  NL,
  SrcBuilder,
  SrcBuilderOptions,
  UNINDENT,
  WS,
} from './src-generator';
import { collectStyles, CssStyles, renderStyles } from './styles';

export type QwikOptions = {
  qwikLib?: string;
  qrlPrefix?: string;
  output?: 'ts' | 'cjs' | 'mjs';
  jsx?: boolean;
  minify?: boolean;
};

export interface FileSet {
  high: File;
  med: File;
  low: File;
}

export function createFileSet(options: QwikOptions = {}): FileSet {
  const opts: Required<QwikOptions> = {
    qwikLib: '@builder.io/qwik',
    qrlPrefix: './',
    output: 'ts',
    minify: false,
    jsx: true,
    ...options,
  };
  const extension =
    (opts.output == 'mjs' ? 'js' : opts.output) + (opts.jsx ? 'x' : '');
  const srcOptions: SrcBuilderOptions = {
    isPretty: !opts.minify,
    isModule: opts.output != 'cjs',
    isTypeScript: opts.output == 'ts',
    isJSX: opts.jsx,
  };
  const fileSet = {
    high: new File(
      'high.' + extension,
      srcOptions,
      opts.qwikLib,
      opts.qrlPrefix,
    ),
    med: new File('med.' + extension, srcOptions, opts.qwikLib, opts.qrlPrefix),
    low: new File('low.' + extension, srcOptions, opts.qwikLib, opts.qrlPrefix),
  };
  Object.defineProperty(fileSet, '_commonStyles', {
    enumerable: false,
    value: { styles: new Map<string, CssStyles>() as any, symbolName: null },
  });
  fileSet.med.exportConst('__merge', minify`${__merge}`);
  return fileSet;
}

function getCommonStyles(
  fileSet: FileSet,
): {
  styles: Map<string, CssStyles>;
  symbolName: string | null;
} {
  return (fileSet as any)['_commonStyles'];
}

export function addComponent(
  fileSet: FileSet,
  component: MitosisComponent,
  opts: { isRoot?: boolean; shareStyles?: boolean } = {},
) {
  const _opts = { isRoot: false, shareStyles: false, ...opts };
  compileAwayBuilderComponentsFromTree(component, {
    ...compileAwayComponents,
    Image: undefined!,
    CoreButton: undefined!,
  });
  const componentName = component.name;
  const handlers = renderHandlers(
    fileSet.high,
    componentName,
    component.children,
  );
  // If the component has no handlers, than it is probably static
  // and so it is unlikely to be re-rendered on the client, therefore
  // put it in a low priority bucket.
  const isStatic = Array.from(handlers.keys()).reduce(
    (p, v) => p && v.indexOf('state') == -1,
    true,
  );
  const onRenderFile = isStatic ? fileSet.low : fileSet.med;
  const componentFile = fileSet.med;
  if (!componentFile.exports.get('onMountCreateEmptyState')) {
    componentFile.exportConst('onMountCreateEmptyState', function(
      this: SrcBuilder,
    ) {
      this.emit(
        invoke(
          componentFile.import(componentFile.qwikModule, 'qHook'),
          [arrowFnValue([], ['({})'])],
          ['any', 'any'],
        ),
      );
    });
  }
  if (component.hooks.onMount?.code) {
    addComponentOnMount(componentFile, componentName, component);
  }
  const styles = _opts.shareStyles
    ? getCommonStyles(fileSet).styles
    : new Map<string, CssStyles>();
  collectStyles(component.children, styles);
  const qComponentOptions: Record<string, any> = {
    // tagName: string(componentName.toLowerCase()),
    onMount: invoke(componentFile.import(componentFile.qwikModule, 'qHook'), [
      componentFile.toQrl(
        component.hooks.onMount?.code
          ? componentName + '_onMount'
          : 'onMountCreateEmptyState',
      ),
    ]),
    onRender: invoke(componentFile.import(componentFile.qwikModule, 'qHook'), [
      onRenderFile.toQrl(componentName + '_onRender'),
    ]),
  };
  if (_opts.shareStyles) {
    if (_opts.isRoot) {
      const symbolName = componentName + '_styles';
      getCommonStyles(fileSet).symbolName = symbolName;
      qComponentOptions.unscopedStyles = onRenderFile.toQrl(symbolName);
    }
  } else {
    if (styles.size) {
      qComponentOptions.styles = onRenderFile.toQrl(componentName + '_styles');
      onRenderFile.exportConst(componentName + '_styles', renderStyles(styles));
    }
  }
  componentFile.exportConst(
    componentName,
    invoke(
      componentFile.import(componentFile.qwikModule, 'qComponent'),
      [qComponentOptions],
      ['any', 'any'],
    ),
  );

  onRenderFile.src.emit(NL);
  const directives: Map<string, string> = new Map();
  onRenderFile.exportConst(
    componentName + '_onRender',
    invoke(onRenderFile.import(onRenderFile.qwikModule, 'qHook'), [
      arrowFnBlock(
        ['__props__', '__state__'],
        [
          renderStateConst(onRenderFile),
          function(this: SrcBuilder) {
            return this.emit(
              'return ',
              renderJSXNodes(
                onRenderFile,
                directives,
                handlers,
                component.children,
                styles,
                {},
              ),
              ';',
            );
          },
        ],
      ),
    ]),
  );
  directives.forEach((code, name) => {
    fileSet.med.import(fileSet.med.qwikModule, 'h');
    fileSet.med.exportConst(name, code);
  });
}

function renderStateConst(file: File, isMount = false): any {
  return function(this: SrcBuilder) {
    return this.emit(
      'const state',
      WS,
      '=',
      WS,
      file.module == 'med'
        ? file.exports.get('__merge')
        : file.import('./med', '__merge').name,
      '(__state__,',
      WS,
      '__props__',
      isMount ? ',true);' : ');',
      NL,
    );
  };
}

export function addCommonStyles(fileSet: FileSet) {
  const { styles, symbolName } = getCommonStyles(fileSet);
  const onRenderFile = fileSet.low;
  if (symbolName) {
    onRenderFile.exportConst(symbolName, renderStyles(styles));
  }
}

function addComponentOnMount(
  componentFile: File,
  componentName: string,
  component: MitosisComponent,
) {
  componentFile.exportConst(componentName + '_onMount', function(
    this: SrcBuilder,
  ) {
    this.emit(
      invoke(
        componentFile.import(componentFile.qwikModule, 'qHook'),
        [
          arrowFnValue(['__props__'], () =>
            this.emit(
              '{',
              NL,
              INDENT,
              'const __state__',
              WS,
              '=',
              WS,
              componentFile.import(componentFile.qwikModule, 'qObject').name,
              '({});',
              NL,
              renderStateConst(componentFile, true),
              iif(component.hooks.onMount?.code),
              NL,
              'return state;',
              UNINDENT,
              NL,
              '}',
            ),
          ),
        ],
        ['any', 'any'],
      ),
    );
  });
}

declare const __STATE__: Record<string, Record<string, any>>;

function __merge(
  state: Record<string, any>,
  props: Record<string, any>,
  create = false,
) {
  for (const key in props) {
    if (
      key.indexOf(':') == -1 &&
      !key.startsWith('__') &&
      Object.prototype.hasOwnProperty.call(props, key)
    ) {
      state[key] = props[key];
    }
  }
  if (create && typeof __STATE__ == 'object' && props.serverStateId) {
    debugger;
    Object.assign(state, __STATE__[props.serverStateId]);
  }
  return state;
}
