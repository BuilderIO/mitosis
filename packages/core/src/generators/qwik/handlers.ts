import { MitosisNode } from '../../types/mitosis-node';
import { renderUseLexicalScope } from './component';
import {
  arrowFnBlock,
  File,
  INDENT,
  invoke,
  NL,
  SrcBuilder,
  UNINDENT,
  WS,
} from './src-generator';

const IIF_START = '(() => {';
const IIF_END = '})()';
export function extractJSBlock(binding: any): string | null {
  if (typeof binding == 'string') {
    if (
      binding.startsWith('{') &&
      binding.endsWith('}') &&
      !binding.startsWith('{"') &&
      !binding.endsWith('"}')
    ) {
      return binding.substring(1, binding.length - 2);
    } else if (binding.startsWith(IIF_START) && binding.endsWith(IIF_END)) {
      return binding.substring(
        IIF_START.length,
        binding.length - IIF_END.length - 1,
      );
    }
  }
  return null;
}

export function renderHandlers(
  file: File,
  componentName: string,
  children: MitosisNode[],
): Map<string, string> {
  let id = 0;
  const map = new Map<string, string>();
  const nodes = [...children];
  while (nodes.length) {
    const node = nodes.shift()!;
    const bindings = node.bindings;
    for (const key in bindings) {
      if (Object.prototype.hasOwnProperty.call(bindings, key)) {
        const binding = bindings[key]!;
        if (binding != null) {
          if (key.startsWith('on')) {
            let block = extractJSBlock(binding) || binding;
            const symbol = `${componentName}_${key}_${id++}`;
            map.set(binding, symbol);
            renderHandler(file, symbol, block);
          }
        }
      }
    }
    nodes.push(...node.children);
  }
  return map;
}

function renderHandler(file: File, symbol: string, code: string) {
  const body = [wrapWithUse(file, code)];
  const shouldRenderStateRestore = code.indexOf('state') !== -1;
  if (shouldRenderStateRestore) {
    body.unshift(renderUseLexicalScope(file));
  }
  file.exportConst(symbol, function (this: SrcBuilder) {
    this.emit([arrowFnBlock([], body)]);
  });
  file.src.emit(NL);
}

function wrapWithUse(
  file: File,
  code: string,
): string | ((this: SrcBuilder) => void) {
  const needsEvent = !!code.match(/\bevent\b/);
  if (needsEvent) {
    return function (this: SrcBuilder) {
      this.emit('{', NL, INDENT);
      needsEvent &&
        this.emit(
          'const event',
          WS,
          '=',
          WS,
          invoke(file.import(file.qwikModule, 'useEvent'), []),
          ';',
          NL,
        );
      const blockContent = stripBlock(code);
      this.emit(blockContent);
      this.emit(UNINDENT, NL, '}');
    };
  } else {
    return code;
  }
}

function stripBlock(block: string) {
  return block.substring(1, block.length - 1).trim();
}
