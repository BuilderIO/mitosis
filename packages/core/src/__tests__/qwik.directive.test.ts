import { Image } from '../generators/qwik/directives';

describe('qwik directives', () => {
  const partial = expect.objectContaining;
  beforeEach(() => {
    (global as any).h = mockH;
  });
  afterEach(() => ((global as any).h = undefined));
  describe('Image', () => {
    test('altText', () => {
      expect(
        Image({ altText: 'foo', image: 'http://some.url' }).children[0][0],
      ).toEqual(
        partial({
          tag: 'img',
          props: partial({ src: 'http://some.url', alt: 'foo' }),
        }),
      );
      expect(Image({ image: 'http://some.url' }).children[0][0]).toEqual(
        partial({
          tag: 'img',
          props: partial({ src: 'http://some.url', role: 'presentation' }),
        }),
      );
    });
    test('builder-pixel is eager', () => {
      expect(
        Image({
          builderBlock: { id: 'builder-pixel-foo' },
          image: 'http://some.url',
        }).children[0][0],
      ).toEqual(
        partial({
          tag: 'img',
          props: partial({ loading: 'eager' }),
        }),
      );
    });
    test('images are lazy loaded', () => {
      expect(
        Image({
          image: 'http://some.url',
        }).children[0][0],
      ).toEqual(
        partial({
          tag: 'img',
          props: partial({ loading: 'lazy' }),
        }),
      );
    });
    test('builder.io URLs are served using webp', () => {
      expect(
        find(
          'source',
          Image({
            image: 'http://foo.builder.io/foo',
          }),
        ),
      ).toEqual(
        partial({
          tag: 'source',
          props: partial({
            srcset: srcSet('http://foo.builder.io/foo?format=webp'),
            type: 'image/webp',
          }),
        }),
      );
    });
    test('maxWidth is added to the srcs', () => {
      expect(
        find(
          'source',
          Image({
            image: 'http://foo.builder.io/foo',
            srcsetSizes: '1234 456',
          }),
        ),
      ).toEqual(
        partial({
          tag: 'source',
          props: partial({
            srcset: srcSet('http://foo.builder.io/foo?format=webp', '1234 456'),
            type: 'image/webp',
          }),
        }),
      );
    });
    test('anchor should wrap in <a>', () => {
      expect(
        find(
          'a',
          Image({
            image: 'http://some.url',
            href: 'my-url',
          }),
        ),
      ).toEqual(
        partial({
          tag: 'a',
          props: partial({ href: 'my-url' }),
        }),
      );
    });
  });
});

function find(tag: string, jsx: MockJSX): MockJSX | undefined {
  if (tag == jsx.tag) return jsx;
  return findInChildren(tag, jsx.children);

  function findInChildren(tag: string, jsxs: MockJSX[]): MockJSX | undefined {
    let value: MockJSX | undefined = undefined;
    for (let i = 0; i < jsxs.length; i++) {
      const jsx = jsxs[i];
      if (Array.isArray(jsx)) {
        value = findInChildren(tag, jsx);
      } else {
        value = find(tag, jsx);
      }
      if (value) return value;
    }
    return undefined;
  }
}

const mockH = (
  tag: string,
  props: Record<string, any>,
  children: any[],
): MockJSX => ({
  tag,
  props,
  children,
});

interface MockJSX {
  tag: string;
  props: Record<string, any>;
  children: MockJSX[];
}

function srcSet(url: string, additionalSizes: string = ''): any {
  return ['100', '200', '400', '800', '1200', '1600', '2000']
    .concat(additionalSizes ? additionalSizes.split(' ') : [])
    .map((size) => {
      const parsedUrl = new URL(url);
      parsedUrl.searchParams.set('width', size);
      return `${parsedUrl} ${size}w`;
    })
    .join(', ');
}
