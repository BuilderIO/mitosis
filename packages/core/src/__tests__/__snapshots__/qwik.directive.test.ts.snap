// Vitest Snapshot v1

exports[`qwik directives > Image > altText 1`] = `
{
  "children": [],
  "props": {
    "alt": "foo",
    "loading": "lazy",
    "src": "http://some.url",
    "style": "object-fit:cover;object-position:center;",
  },
  "tag": "img",
}
`;

exports[`qwik directives > Image > altText 2`] = `
{
  "children": [],
  "props": {
    "loading": "lazy",
    "role": "presentation",
    "src": "http://some.url",
    "style": "object-fit:cover;object-position:center;",
  },
  "tag": "img",
}
`;

exports[`qwik directives > Image > anchor should wrap in <a> 1`] = `
{
  "children": [
    [
      {
        "children": [],
        "props": {
          "loading": "lazy",
          "role": "presentation",
          "src": "http://some.url",
          "style": "object-fit:cover;object-position:center;",
        },
        "tag": "img",
      },
    ],
  ],
  "props": {
    "class": undefined,
    "href": "my-url",
  },
  "tag": "a",
}
`;

exports[`qwik directives > Image > builder.io URLs are served using webp 1`] = `
{
  "children": undefined,
  "props": {
    "srcset": "http://foo.builder.io/foo?format=webp&width=100 100w, http://foo.builder.io/foo?format=webp&width=200 200w, http://foo.builder.io/foo?format=webp&width=400 400w, http://foo.builder.io/foo?format=webp&width=800 800w, http://foo.builder.io/foo?format=webp&width=1200 1200w, http://foo.builder.io/foo?format=webp&width=1600 1600w, http://foo.builder.io/foo?format=webp&width=2000 2000w",
    "type": "image/webp",
  },
  "tag": "source",
}
`;

exports[`qwik directives > Image > builder-pixel is eager 1`] = `
{
  "children": [],
  "props": {
    "loading": "eager",
    "role": "presentation",
    "src": "http://some.url",
    "style": "object-fit:cover;object-position:center;",
  },
  "tag": "img",
}
`;

exports[`qwik directives > Image > images are lazy loaded 1`] = `
{
  "children": [],
  "props": {
    "loading": "lazy",
    "role": "presentation",
    "src": "http://some.url",
    "style": "object-fit:cover;object-position:center;",
  },
  "tag": "img",
}
`;

exports[`qwik directives > Image > maxWidth is added to the srcs 1`] = `
{
  "children": undefined,
  "props": {
    "srcset": "http://foo.builder.io/foo?format=webp&width=100 100w, http://foo.builder.io/foo?format=webp&width=200 200w, http://foo.builder.io/foo?format=webp&width=400 400w, http://foo.builder.io/foo?format=webp&width=800 800w, http://foo.builder.io/foo?format=webp&width=1200 1200w, http://foo.builder.io/foo?format=webp&width=1600 1600w, http://foo.builder.io/foo?format=webp&width=2000 2000w, http://foo.builder.io/foo?format=webp&width=1234 1234w, http://foo.builder.io/foo?format=webp&width=456 456w",
    "type": "image/webp",
  },
  "tag": "source",
}
`;
